src/langcheck/metrics/en/pairwise_text_quality.py

```diff
@@ -12,6 +17,80 @@
 from langcheck.metrics.eval_clients import EvalClient
 from langcheck.metrics.metric_value import MetricValue
 
+from ..prompts._utils import get_template
+
+
+def simulated_annotators(
+    prompt_params: List[dict[str, str | None]],
+    eval_model: EvalClient,
+    k: int = 5,
+    n: int = 5,
+) -> List[float | None]:
+    """Compute a confidence score for the pairwise comparison metric based on
+    the method Simulated Annotators proposed in the paper "Trust or Escalate:
+    LLM Judges with Provable Guarantees for Human Agreement"
+    (https://arxiv.org/abs/2407.18370)
+
+    Args:
+        prompt_params: The parameters used to populate the prompt template.
+        eval_model: The EvalClient instance used for the evaluation.
+        k: the number of examples of preference annotations
+        n: the numbre of simulated annotators
+    Returns:
+        A confidence score for the pairwise comparison metric
+    """
+    # Load preprocessed chatarena data
+    cwd = Path(__file__).parent
+    with open(cwd / "processed_chatarena_examples.jsonl") as f:
+        chatarena_data = [json.loads(line) for line in f]
+    assert len(chatarena_data) >= k, "Not enough examples in the chatarena data"
+
+    # Load the prompt template
+    prompt_template = get_template(
+        "en/confidence_estimating/simulated_annotators.j2"
+    )
+    populated_prompts = [
+        prompt_template.render(prompt_param) for prompt_param in prompt_params
+    ]
+
+    confidence_scores = []
+    for prompt in populated_prompts:
+        # Simulate n annotators
+        prompts = []
+        for _ in range(n):
+            # Generate few-shot examples
+            few_shot_examples = random.sample(chatarena_data, k)
+
+            # Construct the full prompt using k few-shot examples
+            few_shot_prompt = "\n".join(
+                f"[Question]\n{example['prompt']}\n\n"
+                "[Assistant A's response]\n{example['model_a']}\n\n"
+                "[Assistant B's response]\n{example['model_b']}\n\n"
+                "[Verdict]\n{example['winner']}\n"
+                for example in few_shot_examples
+            )
+            prompts.append(
+                re.split(r"\[Few-shot examples\]", prompt)[0]
+                + few_shot_prompt
+                + re.split(r"\[Few-shot examples\]", prompt)[1]
+            )
+
+        # Get the response and log likelihoods
+        responses = eval_model.get_text_responses_with_log_likelihood(prompts)
+        scores = []
+        for response in responses:
+            if response:
+                first_token, log_likelihoods = response[1][0]
+                if first_token in ["A", "B"]:
```

Comment:
I wonder how frequently this is not the case. (I'll try and look into this tonight and get some rough numbers)

Reply:
hmm I tried on the "Why did my parent not invite me to their wedding?" example and 4/5 were not A or B

<img width="977" alt="image" src="https://github.com/user-attachments/assets/9382f7e9-5513-4cd1-83a5-d7356a0f7d0f">

not sure what we should do about this 🤔 we can discuss tmrw!

Reply:
(Dupe of another thread. Resolving)